### PR TITLE
Add the ability to use multiple labels as filters

### PR DESCRIPTION
**Pull request by erjiaqing**
Implements Issue #3430 

multiplelabels

As in GitHub, when selecting multiple labels, only display those issues with all the labels selected.

I cannot figure out how to achieve this using the orm provided, so I write a subquery.

Then I use `Label.QueryString` and `Label.IsSelected` for the query string(issues?labels=xxx) and the tick before labels.

### CODE REVIEW

File: models/issue_label.go [hunk 1]
```diff
@@ -76,6 +78,25 @@ func (label *Label) CalOpenIssues() {
 	label.NumOpenIssues = label.NumIssues - label.NumClosedIssues
 }
 
+// CalQueryString calculates query string in issue/pulls list
+func (label *Label) CalQueryString(query []string) {
```
Comment (jonasfranz) [inline models/issue_label.go]: You should rename `query []string` to `labels []string`.
I think that the function name is not very self-explanatory at the first place. It should be renamed into `RestoreSelectedLabels` or `LoadSelectedLabelsFromQuery`.

**Review by jonasfranz — CHANGES_REQUESTED** [review]

**Comment by erjiaqing** [comment]
In GitHub, when multiple labels are selected, only issues with **all of** the selected labels will be shown. However, in test, it is assumed that issues with **at least one of** the selected labels will be shown.
Maybe I have to change the tests?

**Comment by ernierasta** [comment]
Drone will not show, what is the problem with CI... so I have no idea how serious problem with this commit is ...

This would be life saver, is there reason why it is not merged?
Can we make it merge? ;-)

**Comment by erjiaqing** [comment]
I could remember that the last time, it reports that my commit cannot pass the unit test
it is because that:
```
 			IssuesOptions{
 				Labels:   "1,2",
 				Page:     1,
 				PageSize: 4,
 			},
-			[]int64{5, 2, 1},
+			[]int64{1}, // issues with **both** label 1 and 2, only issue 1 matches
 		},
```

I made a net  commit, and it should pass the test now, and this time, it reports another failure
I have no idea about this, maybe the structure of database had changed a lot
@ernierasta 
(I don't know if you will be notified for this)

File: models/issue.go [hunk 2]
```diff
@@ -1140,10 +1140,20 @@ func (opts *IssuesOptions) setupSession(sess *xorm.Session) error {
...
 		}
-		if len(labelIDs) > 0 {
+		if len(labelIDs) == 1 {
 			sess.
 				Join("INNER", "issue_label", "issue.id = issue_label.issue_id").
 				In("issue_label.label_id", labelIDs)
+		} else if len(labelIDs) > 1 {
+			cond, args, _ := builder.ToSQL(builder.In("issue_label.label_id", labelIDs))
+			sess.
+				Where(fmt.Sprintf(`issue.id IN (
```
Comment (lafriks) [inline models/issue.go:1150]: Can't this be built using builder? cc: @lunny

**Comment by ghost** [comment]
https://github.com/erjiaqing/gitea/blob/5f08f1bc0ac685c8e0fd983ca652ee363625395f/templates/repo/issue/list.tmpl#L135
```
<span class="octicon {{if eq $.SelectLabels .ID}}octicon-check{{end}}"></span><span class="label color" style="background-color: {{.Color}}"></span> {{.Name}}
```
`eq $.SelectLabels .ID` -> `.IsSelected`?

https://github.com/erjiaqing/gitea/blob/5f08f1bc0ac685c8e0fd983ca652ee363625395f/routers/repo/issue.go#L243
```
ctx.Data["SelectLabels"] = com.StrTo(selectLabels).MustInt64()
```
`selectLabels` is no longer an int

File: models/issue.go [hunk 3]
```diff
@@ -1322,9 +1332,19 @@ func GetIssueStats(opts *IssueStatsOptions) (*IssueStats, error) {
 			labelIDs, err := base.StringsToInt64s(strings.Split(opts.Labels, ","))
 			if err != nil {
 				log.Warn("Malformed Labels argument: %s", opts.Labels)
-			} else if len(labelIDs) > 0 {
+			} else if len(labelIDs) == 1 {
```
Comment (bkcsoft) [inline models/issue.go:1335]: This query could be fixed to do `> 0` instead of having 2 different queries

**Pull request closed without merging**